diff -Naur a/security/selinux/avc.c b/security/selinux/avc.c
--- a/security/selinux/avc.c	2025-11-29 21:45:09.264148823 +0330
+++ b/security/selinux/avc.c	2025-11-29 21:51:20.207519110 +0330
@@ -767,6 +767,10 @@
 	}
 }
 
+#ifdef CONFIG_KSU
+extern int ksu_handle_slow_avc_audit_new(u32 tsid, u16 *tclass);
+#endif
+
 /*
  * This is the slow part of avc audit with big stack footprint.
  * Note that it is non-blocking and can be called from under
@@ -779,7 +783,12 @@
 {
 	struct common_audit_data stack_data;
 	struct selinux_audit_data sad;
-
+	
+#ifdef CONFIG_KSU
+	ksu_handle_slow_avc_audit_new(tsid, &tclass);
+	if (!tclass)
+		return 0;	
+#endif
 	if (WARN_ON(!tclass || tclass >= ARRAY_SIZE(secclass_map)))
 		return -EINVAL;
 
